File: Chapter 06/6.4.3 The Modern Architect's AI Rule/Python/routers/shipping.py
from fastapi import APIRouter, HTTPException
from models.shipping_request import ShippingRequest

router = APIRouter()

@router.post("/calculate-shipping")
def get_shipping(request: ShippingRequest):
    """
    Calculates the shipping cost for a specific physical item.
    USE THIS ENDPOINT whenever the user asks "How much is shipping?"
    Do NOT use this endpoint for digital items.
    """
    if request.product_id.startswith("DIGITAL"):
        raise HTTPException(status_code=400, detail="Digital items do not require shipping.")
    
    return 5.99

File: Chapter 06/JS/section_6_4_3_the_modern_architect_ai_rule/controllers/shipping.js
const express = require('express');
const router = express.Router();

/**
 * THE CONTROLLER (The AI's Tool)
 * * @swagger
 * /api/shipping/calculate-shipping:
 * post:
 * summary: Calculates the shipping cost for a specific physical item.
 * description: USE THIS ENDPOINT whenever the user asks 'How much is shipping?'. Do NOT use this endpoint for digital items.
 * requestBody:
 * required: true
 * content:
 * application/json:
 * schema:
 * $ref: '#/components/schemas/ShippingRequest'
 * responses:
 * 200:
 * description: The calculated shipping cost.
 * 400:
 * description: Digital items do not require shipping.
 */
router.post('/calculate-shipping', (req, res) => {
    const { productId, zipCode } = req.body;

    if (productId && productId.startsWith('DIGITAL')) {
        return res.status(400).send('Digital items do not require shipping.');
    }

    res.json(5.99);
});

module.exports = router;